### ISSUE TITLE
Add typescript `.d.ts` bindings

### ISSUE THREAD

**Issue by gkjohnson** [gkjohnson/three-mesh-bvh#79]

**Comment by crabmusket** [comment]
I'm happy to take ownership of this one. What's your preferred strategy for dealing with types? I assume just adding a `src/index.d.ts` file covering the main exports, making sure to be compatible with THREE's core types? Do you have any thoughts on testing types?

**Comment by gkjohnson** [comment]
Excellent! Thank you for the offer.

> I assume just adding a src/index.d.ts file covering the main exports, making sure to be compatible with THREE's core types

That sounds like a good place to start. In other projects I've created d.ts sibling files for everything that was part of the public interface but I think the interface is small enough here that it's manageable in one file. There should be a separate types file for `src/workers/GenerateMeshBVHWorker.js`, though, which isn't currently exported from index.js because Workers are not compatible with most build processes by default.

> Do you have any thoughts on testing types?

I am not a typescript expert so I'm open to suggestions for validating types but on other projects I've just added a `tsc` command to the lint script in the package.json with a `noEmit` ts config file to make sure all the syntax is valid.

And you'll know better than me but if there's anything special that needs to be done in order to enable [the prototype modification](https://github.com/gkjohnson/three-mesh-bvh#use) that the docs suggest or setting custom parameters on existing three.js types such as [Raycaster.firstHitOnly](https://github.com/gkjohnson/three-mesh-bvh#raycasterfirsthitonly) it might be nice to add documentation for ts-specific use. Of course not required and anything you can contribute is appreciated!

**Comment by crabmusket** [comment]
Adding a `noEmit` compiler check sounds like a good way to get started with tests. I'm no expert either! I think the only challenge will be staying compatible with THREE itself as it progresses through versions.

The prototype modification shouldn't be an issue, but I'll investigate whether it's possible to modify the `Raycaster` type when this library is installed.

I'll see what I can do over the weekend :)

**Comment by crabmusket** [comment]
Working branch: https://github.com/gkjohnson/three-mesh-bvh/compare/master...crabmusket:typechecking

**Comment by gkjohnson** [comment]
Awesome! It's looking good. I didn't know about the `declare module` syntax for modifying existing type definitions -- seems useful

**Comment by gkjohnson** [comment]
Hey @crabmusket what are your plans for the type definitions? No rush I was just planning to merge the v0.4.0 (#205) possibly in the next few weeks which introduces some API adjustments so I didn't want to make changes out from under you if you were planning to make a PR!

**Comment by crabmusket** [comment]
Go ahead and merge, I've had an extra busy time at work so haven't been able to progress on this for a bit. I'll adapt once those changes are merged :)

**Comment by Mathias-Fuchs** [comment]
In case anyone trying to make raycasts work from Typescript, here's a simple and quick solution that worked for me:
```typescript
THREE.Mesh.prototype.raycast = acceleratedRaycast;
class MBG extends THREE.BufferGeometry {
    boundsTree: any;
    constructor(bg: THREE.BufferGeometry) {
        super();
        this.copy(bg);
        this.boundsTree = new MeshBVH(this as THREE.BufferGeometry);
    };
}
```
instead of using Typescript's aweful "declare module" -syntax, and then, for instance:
```typescript
        const tkbg = new THREE.TorusKnotBufferGeometry();
        const tkbgm = new MBG(tkbg);
        const testmesh = new THREE.Mesh(tkbgm, new THREE.MeshBasicMaterial({ color: "red", wireframe: true }));
        scene.add(testmesh);
```
and then one can testmesh.raycast(...) as usual, it will pick up the attached boundsTree.

**Comment by gkjohnson** [comment]
Fixed in #338

**Comment by clicktodev** [comment]
sure i'll open a pr for jsdoc type generation
